dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+    period, one is set immediately after the other
+* Note the lack of logger here. See "Logging" later, but in general .NET doesn't have a logger in the standard library.
+  We may make some kind of `IReadOnlyDictionary<string, string> ActivityInfo.LogInfo { get; }` that can be used with
+  `ILogger.BeginScope` or similar.
+
+### Async Completion
+
+Simply `throw new Temporal.Activity.CompleteAsyncException` from the activity.
+
+## Client
+
+Inside the `Temporal.Client` namespace, we have the following types:
+
+* `TemporalClient` - namespace-specific, stateless high-level client to Temporal
+  * See later for most of its API design
+  * Accepts `ITemporalConnection`
+  * Implements `ITemporalClient` for DI use (which itself implements `Temporal.Worker.IWorkerClient` for worker use)
+* `TemporalConnection` - namespace-agnostic, stateful low-level client to Temporal
+  * Has `WorkflowService`, `OperatorService`, etc for making raw gRPC calls
+  * Supports lazy connectivity
+  * Implements `ITemporalConnection` for DI use
+* `WorkflowHandle` - high-level workflow-specific calls
+* Many other supporting types
+
+Example of starting the workflow defined earlier in document, signalling it, then waiting on its results:
+
+```csharp
+using Temporal.Client;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Start workflow
+var handle = await client.StartWorkflowAsync(
+    IGreetingWorkflow.Ref.RunAsync, "Temporal", new(ID: "my-workflow-id", TaskQueue: "my-task-queue"));
+
+// Send signal
+await handle.Signal(IGreetingWorkflow.Ref.CompleteWithGreetingAsync);
+
+// Wait for result
+Console.WriteLine(await handle.GetResultAsync());
+```
+
+Example of doing the same via already started workflow:
+
+```csharp
+using Temporal.Client;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Get workflow handle
+var handle = await client.GetWorkflowHandle<string>("my-workflow-id");
+
+// Send signal
+await handle.Signal(IGreetingWorkflow.Ref.CompleteWithGreetingAsync);
+
+// Wait for result
+Console.WriteLine(await handle.GetResultAsync());
+```
+
+Note, this could have been just `client.GetWorkflowHandle("my-workflow-id")` + `handle.GetResultAsync<string>()` if the
+user would rather not provide the type until fetching the result. See the definitions below.
+
+### TemporalClient
+
+Looks like:
+
+```csharp
+namespace Temporal.Client;
+
+public class TemporalClient : ITemporalClient
+{
+    public static async Task<TemporalClient> ConnectAsync(TemporalClientConnectOptions options) { /*...*/ }
+
+    public TemporalClient(ITemporalConnection connection, TemporalClientOptions options) { /*...*/ }
+
+    public string Namespace { get; }
+    public DataConverter DataConverter { get; }
+    public IBridgeClientProvider BridgeClientProvider { get; }
```

Comment:
Do you really intend to expose this? Why?

Reply:
I do. I have a similar thing in Python.

My reasoning here is that people want to make their own clients or wrap their own clients, but we still need this usable from a worker. We messed this up in Go and now people can't wrap clients the way they want because we cast back to the concrete type when using in a worker. So we have this interface that `ITemporalClient` implements (left out for brevity):

```csharp
namespace Temporal.Worker;

public interface IWorkerClient
{
    public string Namespace { get; }
    public Temporal.Converters.DataConverter DataConverter { get; }
    public Temporal.Client.IBridgeClientProvider BridgeClientProvider { get; }
}
```

So if you wrap your client, so long as it still implements `IWorkerClient` (which is implemented by `ITemporalClient` which is probably what you'd conform to), you're ok. So you'd just pass through the bridge client provider from the underlying client.

I would definitely entertain better options here. The goal is "Allow easy wrapping/impl of the Temporal client interface while still allowing that to be used to create workers".

Reply:
>My reasoning here is that people want to make their own clients or wrap their own clients, but we still need this usable from a worker. We messed this up in Go and now people can't wrap clients the way they want because we cast back to the concrete type when using in a worker.

I get it.

Could we have a single `asWorkerClient()` method, that returns an opaque `WorkerClient` instead? The key difference here would be that this object would be guaranteed to be immutable (we don't have that certainty on a user implementation of `IWorkerClient`) and that `BridgeClientProvider` would not be publicly exposed.

I can see why devs might want to extends `TemporalClient`, but I'd prefer not provide them a footgun if we can avoid that...